### PR TITLE
Improved expanding source paths of volumes

### LINKED ISSUE #6233: Expanding volumes source paths
## Description of the issue
When volume defined with long syntax, its source paths which starts with `~` do not expand well as a paths started with `.`
## Context information (for bug reports)

```
docker-compose version 1.22.0, build f46880f
```

```
version: "3.4"

services:
  test:
    image: "test:${DISTRO:-trusty}"
    build:
      context: test/
      args: { version: "${DISTRO:-trusty}" }
    volumes:
      - type: bind
        read_only: true
        source: ~/tmp/test1
        target: /tmp/test1
      
      - ~/tmp/test2:/tmp/test2:ro
```

## Steps to reproduce the issue

1. Define volume with use of long syntax and set source started with `~`
2. `docker-compose up -d --force-recreate <service>`

### Observed result
```
Recreating test_test_1 ... error

ERROR: for test_test_1  Cannot create container for service test: invalid mount config for type "bind": invalid mount path: '~/tmp/test1' mount path must be absolute

ERROR: for test  Cannot create container for service test: invalid mount config for type "bind": invalid mount path: '~/tmp/test1' mount path must be absolute
ERROR: Encountered errors while bringing up the project.
```

### Expected result
```
Recreating 48d9f6c670d9_test_test_1 ... done
```

## Additional information

`Ubuntu 18.04.1 LTS`

### PR DESCRIPTION
**Pull request by zasca**
defined with long syntax when paths starts with '~'
Resolves #6233

### CODE REVIEW

**Comment by shin-** [comment]
Thanks! Can you try and add a test for this as well?

**Comment by zasca** [comment]
If my unit test is correct, then path expanding do not properly work on Windows platform and need more complex patching.

File: tests/unit/config/config_test.py [hunk 1]
```diff
@@ -1322,6 +1322,36 @@ def test_load_bind_mount_relative_path(self):
         assert mount.type == 'bind'
         assert mount.source == expected_source
 
+    @mock.patch.dict(os.environ)
+    def test_load_bind_mount_relative_path_with_tilda(self):
+        if IS_WINDOWS_PLATFORM:
```
Comment (shin-) [inline tests/unit/config/config_test.py]: I think we can do without that - assume that Python expands `~` correctly on all platforms, and change the test to verify that `mount.source` doesn't start with `~` and has `/web` as the last element of its path.

Comment (zasca) [reply tests/unit/config/config_test.py]: I updated the test as you suggest, but the result is the same.

**Comment by zasca** [comment]
Now everything looks fine.

File: tests/unit/config/config_test.py [hunk 2]
```diff
@@ -1322,6 +1322,29 @@ def test_load_bind_mount_relative_path(self):
...
+            },
+        )
+
+        details = config.ConfigDetails('.', [base_file])
+        config_data = config.load(details)
+        mount = config_data.services[0].get('volumes')[0]
+        assert mount.target == '/web'
+        assert mount.type == 'bind'
+        assert (not mount.source.startswith('~')
+                and mount.source.endswith('\\web'))
```
Comment (shin-) [inline tests/unit/config/config_test.py]: That's going to fail on UNIX systems. Use `endswith('{}web'.format(os.path.sep))` instead?

File: tests/unit/config/config_test.py [hunk 3]
```diff
@@ -1322,6 +1322,29 @@ def test_load_bind_mount_relative_path(self):
         assert mount.type == 'bind'
         assert mount.source == expected_source
 
+    def test_load_bind_mount_relative_path_with_tilda(self):
```
Comment (shin-) [inline tests/unit/config/config_test.py]: nit: tild**e**

File: tests/unit/config/config_test.py (hunk 2 above)
Comment (zasca) [reply tests/unit/config/config_test.py]: Your truth. Thanks.

File: tests/unit/config/config_test.py (hunk 3 above)
Comment (zasca) [reply tests/unit/config/config_test.py]: Fixed.

**Comment by zasca** [comment]
This PR with last changes is good for merging now?

**Comment by zasca** [comment]
> @zasca Looks good now yeah - can you squash this down to a single commit?

Done

**Comment by shin-** [comment]
Thanks!

**Pull request merged**